pkg/model/secret/secret.go

```diff
@@ -47,19 +49,81 @@ type KubernetesSecret struct {
 
 type SecretLoader interface {
 	Load(secret *Secret) (string, error)
+	Mount(secret *Secret) (string, error)
 }
 
 type secretLoader struct {
 	// secretLoader is limited to a single namespace, to avoid hijacking other namespace's secrets
-	namespace string
-	client    client.Reader
+	namespace   string
+	client      client.Client
+	mountConfig *MountConfig
 }
 
-func NewSecretLoader(client client.Reader, namespace string) *secretLoader {
+// MountConfig to attach volume secrets to Fluentd
+type MountConfig struct {
+	SecretName      string
+	SecretNamespace string
+	ConfigPath      string
+	LoggingRef      string
+}
+
+func NewSecretLoader(client client.Client, namespace string, mountConfig *MountConfig) *secretLoader {
 	return &secretLoader{
-		client:    client,
-		namespace: namespace,
+		client:      client,
+		namespace:   namespace,
+		mountConfig: mountConfig,
+	}
+}
+
+func (k *secretLoader) Mount(secret *Secret) (string, error) {
+	k8sSecret := &corev1.Secret{}
+	err := k.client.Get(context.TODO(), types.NamespacedName{
+		Name:      secret.MountFrom.SecretKeyRef.Name,
+		Namespace: k.namespace}, k8sSecret)
+	if err != nil {
+		return "", errors.WrapIff(err, "failed to get kubernetes secret %s:%s",
+			k.namespace,
+			secret.MountFrom.SecretKeyRef.Name)
+	}
+	secretKey := fmt.Sprintf("%s-%s-%s", k.namespace, secret.MountFrom.SecretKeyRef.Name, secret.MountFrom.SecretKeyRef.Key)
+	value, ok := k8sSecret.Data[secret.MountFrom.SecretKeyRef.Key]
+	if !ok {
+		return "", errors.Errorf("key %q not found in secret %q in namespace %q",
+			secret.MountFrom.SecretKeyRef.Key,
+			secret.MountFrom.SecretKeyRef.Name,
+			k.namespace)
+	}
+	fluentOutputSecret := &corev1.Secret{}
+	err = k.client.Get(context.TODO(), types.NamespacedName{
+		Name:      k.mountConfig.SecretName,
+		Namespace: k.mountConfig.SecretNamespace}, fluentOutputSecret)
+	if err != nil {
+		return "", errors.Wrap(err, fmt.Sprintf("unable to get fluent output secret: %q", k.mountConfig.SecretName))
+	}
+	if fluentOutputSecret.Data == nil {
+		fluentOutputSecret.Data = make(map[string][]byte)
+	}
+	fluentOutputSecret.Data[secretKey] = value
+	err = k.client.Update(context.TODO(), fluentOutputSecret)
+	if err != nil {
+		return "", errors.Wrap(err, fmt.Sprintf("unable to update fluent output secret: %q", k.mountConfig.SecretName))
+	}
+	var loggingRef string
+	if k.mountConfig.LoggingRef != "" {
+		loggingRef = k.mountConfig.LoggingRef
+	} else {
+		loggingRef = "default"
+	}
+	annotationKey := fmt.Sprintf("logging.banzaicloud.io/%s", loggingRef)
+	if k8sSecret.ObjectMeta.Annotations == nil {
+		k8sSecret.ObjectMeta.Annotations = make(map[string]string)
+	}
+	k8sSecret.ObjectMeta.Annotations[annotationKey] = "watched"
+	err = k.client.Update(context.TODO(), k8sSecret)
+	if err != nil {
+		return "", errors.Wrap(err, fmt.Sprintf("unable to update source secret: %q", secret.MountFrom.SecretKeyRef.Name))
 	}
+	return k.mountConfig.ConfigPath + "/" + secretKey, nil
```

Comment:
I think there are some issues with this: we update all the source/target secrets all the time, which will cause a lot of event updates and also it could cause several fluentd reloads as well. The loader should not have any side effects, all the secret preparation should happen in the main reconcile loop. We should move the logic there if possible and also be careful to avoid updating resources if they haven't changed.

Reply:
This ha been refactored. The new approach only collects secrets during config generation and reconcile it before the other reconcile cycles.